### PR TITLE
async client: add hash_policy option

### LINKED ISSUE #4899: AsyncClient returned by httpAsyncClientForCluster() does not follow lb_policy defined for cluster
*Title*: *AsyncClient returned by httpAsyncClientForCluster() does not follow lb_policy defined for cluster*

*Symptom*:
LB Policy configured on a test cluster is RING_HASH, and I observed it was honoured by all calls going through HTTP Connection manager.
But any Async call sent using AsyncClient returned by httpAsyncClientForCluster() from a filter is **not following the same LB policy**. Is this the expected behavior?

*Description*:

Here is what is causing the problem:

In file `upstream/cluster_manager_impl.cc` the constructor of `struct ClusterEntry` is called for each cluster defined in the config when envoy starts up.

 In its initializer list it invokes the constructor of `Http::AsyncClientImpl` to create http_async_client_ for the cluster. http_async_client_ is returned by httpAsyncClientForCluster()  and is used for making Async calls. 

`Http::AsyncClientImpl` creates `AsyncStreamImpl` which creates a mock `Router::ProdFilter router_` and sets a bunch of mock objects on it. One of the mock objects is `struct RouteEntryImpl` which defines `const Router::HashPolicy* hashPolicy() const override { return nullptr; }`.

*During an async call*, when `upstream/thread_aware_lb_impl.cc` chooseHost()  is called, it tries to `computeHashKey()` but in the absence a HashPolicy, it does this:

`const uint64_t h = hash ? hash.value() : random_.random();`

Do you think there a way to make async calls start following the lb_policy for the cluster?

### PR DESCRIPTION
**Pull request by kyessenov**
Signed-off-by: Kuat Yessenov <kuat@google.com>

Description:
Add hash policy to the async client. This is needed to support consistent hash load balancing for external requests originating from the proxy (e.g. telemetry, ext authz, WASM, etc).
PR moves hash policy implementation to http namespace with no functional changes.

/cc @mandarjog 

Risk Level: low
Testing: unit test
Docs Changes: none
Release Notes: none
Fixes #4899

### CODE REVIEW

**Comment by kyessenov** [comment]
Friendly ping.

**Comment by kyessenov** [comment]
Ping?

**Comment by junr03** [comment]
Whoops sorry, this one fell through the cracks last week. I'll review by tomorrow.

**Comment by mattklein123** [comment]
@junr03 @goaway please review. Thank you.

File: include/envoy/http/async_client.h [hunk 1]
```diff
@@ -199,6 +210,11 @@ class AsyncClient {
       StreamOptions::setSendXff(v);
       return *this;
     }
+    RequestOptions& setHashPolicy(
+        const Protobuf::RepeatedPtrField<envoy::api::v2::route::RouteAction::HashPolicy>& v) {
```
[the remark this answers is not in the document]
Comment (kyessenov) [reply include/envoy/http/async_client.h]: I think it needs to downcast to request options, so requires an override in a subclass. It's a config options class AFAICT.

File: include/envoy/http/async_client.h [hunk 2]
```diff
@@ -199,6 +210,11 @@ class AsyncClient {
...
+      hash_policy = v;
```
Comment (junr03) [inline include/envoy/http/async_client.h]:
```suggestion
      StreamOptions::setHashPolicy(v);
```

Sorry about the confusion above. I meant to tag this line.

Comment (kyessenov) [reply include/envoy/http/async_client.h]: Ah, got it. Yes, updated, thanks.

**Comment by kyessenov** [comment]
The failing test looks like a random chance (birthday paradox?)
Can someone retrigger it?

**Comment by kyessenov** [comment]
This is finally de-flaked!
@junr03 do you mind taking a look?

**Comment by lizan** [comment]
LGTM, can you merge master to pick latest CI change?

**Comment by kyessenov** [comment]
@lizan Ping?

**Comment by lizan** [comment]
@kyessenov can you merge master?

**Comment by kyessenov** [comment]
@lizan Done master merge.

**Pull request merged**